### PR TITLE
RSS Feed Subscription Integration

### PR DESCRIPTION
**Pull request by flamingquaks**
# RSS Subscription Integration
This Feature introduces the ability for a RAG workspace to subscribe to an RSS Feed, check for new posts daily and send new posts for Website Crawling for the RAG Workspace. This can showcase simple automation for consuming data for RAG / Semantic Search.

Screenshot 2023-11-17 at 16 16 28

## How it Works
Users can subscribe to a RSS Feed by navigating to a desired RAG workspace and navigating to the "RSS Feeds" tab on the workspace documents table. 

Clicking the "Subscribe to RSS Feed" button on the table or by clicking "Add RSS Subscription" in the "Add Data" drop down will navigate the user to the RAG Workspace's "Add Data" view, with the RSS Feeds Tab selected. 
The user can paste a link to an RSS Feed and provide a title for the feed. 

Screenshot 2023-11-22 at 18 21 03

Once the user successfully subscribes to the RSS Feed, the feed will be visible in the "RSS Feeds" tab of the RAG Workspace documents table. 
The RSS Feed is then sent to be checked for posts, which are then added to the RSS Feed DynamoDB table.

An independent scheduled lambda checks the RSS Feed table for posts that are in a pending status and sends up to 10 pending posts for to be Website Crawled into the RAG Workspace. The RSS Feed Check & Post Crawling were designed to be decoupled to provide a fixed rate-limit for the number of websites being crawled. Additionally, the architecture leverages an EventBridge schedule for each post with a flexible timeframe for execution. This can also help distribute work over time if the user has subscribed to a high-volume of feeds or feeds that have a high-volume of posts. Since the posts are ingested using the existing crawl websites functionality, the crawled blog posts will show under the websites once they are consumed. 

Additionally, a user can click on a RSS Feed subscription from the documents table on the RAG workspace, which takes the user to a RSS Feed details page. The page gives the details of the RSS Feed subscription and the posts that have been found from the feed. It also provides the user with the ability to enable/disable a subscription. 

Screenshot 2023-11-22 at 18 19 11
Screenshot 2023-11-22 at 18 20 35

Finally, if a user deletes a RAG workspace, the RSS Subscription's EventBridge schedule is deleted and any RSS Feed or Post data relating to the workspace is removed from the Documents DynamoDB table.

Nov-17-2023 16-40-20

## High-Level Technical Changes/Additions
The goal with changes was to leverage as much of the existing design patterns/structures. 
* 3 New lambdas are added. 
* 1. The first lambda runs once daily by EventBridge rule. This lambda pulls all RSS Feeds that are currently enabled across any workspace. It loops through each and Invokes an RSS Ingestor Lambda function.
* 1. The second lambda is triggered by the previous lambda. This lambda will parse the RSS feed for the provided `workspace_id`/`feed_id`. The lambda will add newly detected posts to the documents table as `rsspost` type documents. The documents have a status of `pending`. If a post already exists, the lambda will not overwrite it to preserve the original data and ensure process RSS feed posts aren't re-processed.
* 1. The third lambda runs every 10 minutes. This lambda queries the documents table for documents with a status of `pending` and a document type of `rsspost`. It then submits each post for website crawling. The website crawling will be associated with the workspace the RSS feed/posts exist.

The design of the three lambda architecture was based on a few factors:

* The first lambda, which triggers one or more of the second lambda was designed to be the single RSS post check trigger, keeping scheduling simple. The decision to have each RSS feed execute feed crawling in an isolated lambda was to ensure scalability. While feed parsing is relatively quick, if users subscribe to a high volume of feeds, an execution time out could occur. 
* With the second lambda executing feed parsing on one feed per execution, it can help mitigate failures due DDB capacity issues (in high volume scenarios) by allowing smaller retries w/ exponential back off. Additionally, debugging and tracing is easier with isolated feed parsing executions.
* The third lambda is designed to ensure the Website Crawling SageMaker Inference endpoint isn't overwhelmed. Leveraging DDB Documents Table as a processing queue, it grabs a max of 10 websites every 10 minutes. 

This PR is relating to #226.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -38,7 +38,8 @@
     "@aws-cdk/aws-apigatewayv2-authorizers-alpha": "^2.104.0-alpha.0",
     "@aws-cdk/aws-apigatewayv2-integrations-alpha": "^2.104.0-alpha.0",
     "@aws-cdk/aws-cognito-identitypool-alpha": "^2.104.0-alpha.0",
-    "aws-cdk-lib": "2.104.0",
+    "@aws-solutions-constructs/aws-eventbridge-lambda": "^2.46.0",
+    "aws-cdk-lib": "2.105.0",
```
Comment (bigadsoleiman) [inline package.json:42]: We want to be consistent on versioning and we usually bump on dedicated PR after testing deployments of the all infra with all components before doing so

Any specific reason to bump to 2.105 here? can this be left at 2.104 for this PR?

**Review by bigadsoleiman — CHANGES_REQUESTED** [review]

**Comment by bigadsoleiman** [comment]
Great contribution @flamingquaks! looking into it

File: package.json (hunk 1 above)
Comment (flamingquaks) [reply package.json:42]: @bigadsoleiman I agree with you about version changes. 
I introduced the AWS solutions construct which had peer dependencies on v2.105. 
Alternatively, I can remove the construct and simply create an event bridge rule manually. There's little effort to make that change and I can revert the version bump.

**Comment by flamingquaks** [comment]
@bigadsoleiman thanks!
I'm hoping to have finally tweaks done today. I had some over-engineering introduced which I'm yanking out. The final push should be today and I'll take the PR out of draft.

**Comment by flamingquaks** [comment]
I've moved the PR back to ready!

From the original PR state to now, I pulled out the `rss` specific logic and leveraged `document` logic where possible, adding handling by the `document_type` of either `rssfeed` or `rsspost` depending on the case. 

I've also simplified the scheduling/feed parsing/crawl queuing, which I updated in the PR description. 

Feature-wise, I've also removed the original hard-coded "follow links" and "link limit" parameters the RSS logic was using when sending a website for crawling. The user can now, in the UI, set follow links and limit, just like on the Add Website to Crawl page. 
On the RSS Feed details page, the user can toggle the RSS subscription on/off. The user can also modify the "follow links" and "link limit" values applied to the RSS feed, which will apply the settings to any newly found RSS feed posts.

The original requirement to bump the `aws-cdk-lib` version has been removed as I removed the `aws-solutions-construct` I was leveraging in the CDK stack. 

Within the `requirements.txt` there's an added dependency for `feedparser` https://pypi.org/project/feedparser/

File: lib/rag-engines/data-import/functions/trigger-rss-ingestors/index.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+
+@tracer.capture_lambda_handler()
+@logger.inject_lambda_context(log_event=True)
+def lambda_handler(event, context: LambdaContext):
+    logger.info(f"Triggering daily checks for RSS Feed Posts")
+    try:
+        genai_core.documents.ingest_rss_feeds()
+    except Exception as e:
+        logger.error(f'Error triggering RSS Feed checks')
```
Comment (massi-ang) [inline lib/rag-engines/data-import/functions/trigger-rss-ingestors/index.py:18]: remove `f` since the string is not a template

File: lib/rag-engines/data-import/functions/rss-ingestor/index.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+def lambda_handler(event, context: LambdaContext):
+    logger.info(f"Starting scheduled RSS Feed poll")
+    workspace_id = event['workspace_id']
+    document_id = event['document_id']
+    logger.info(f"workspace_id = {workspace_id}")
+    logger.info(f'document_id = {document_id}')
+    try:
+        genai_core.documents.check_rss_feed_for_posts(workspace_id,document_id)
+    except Exception as e:
+        logger.error(f'Error checking for new posts from feed!')
```
Comment (massi-ang) [inline lib/rag-engines/data-import/functions/rss-ingestor/index.py:22]: remove `f` since the string is not a template

File: lib/rag-engines/data-import/website-crawling-workflow.ts
Comment (massi-ang) [inline lib/rag-engines/data-import/website-crawling-workflow.ts:1]: Wouldn't be better to have a separate construct for the RSS Ingestor?

File: lib/rag-engines/workspaces/delete-workspace.ts
Comment (massi-ang) [inline lib/rag-engines/workspaces/delete-workspace.ts:1]: run `prettify` formatter

File: lib/shared/layers/python-sdk/python/genai_core/documents.py [hunk 4]
```diff
@@ -23,12 +27,19 @@
     "DEFAULT_KENDRA_S3_DATA_SOURCE_BUCKET_NAME"
 )
 
+RSS_FEED_INGESTOR_FUNCTION = os.environ['RSS_FEED_INGESTOR_FUNCTION'] if "RSS_FEED_INGESTOR_FUNCTION" in os.environ else ""
```
Comment (massi-ang) [inline lib/shared/layers/python-sdk/python/genai_core/documents.py:30]: use `os.environ.get("VAR_NAME", "")`

File: lib/shared/layers/python-sdk/python/genai_core/documents.py [hunk 5]
```diff
@@ -23,12 +27,19 @@
...
+RSS_FEED_SCHEDULE_ROLE_ARN = os.environ['RSS_FEED_SCHEDULE_ROLE_ARN'] if "RSS_FEED_SCHEDULE_ROLE_ARN" in os.environ else ""
```
Comment (massi-ang) [inline lib/shared/layers/python-sdk/python/genai_core/documents.py:31]: same as above

File: lib/shared/layers/python-sdk/python/genai_core/documents.py [hunk 6]
```diff
@@ -48,6 +60,11 @@ def list_documents(
     if document_type == "text" or document_type == "qna":
         scan_index_forward = False
 
+    sort_key_prefix = f"{document_type}/"
+    if document_type == "rsspost" and parent_document_id != None:
+        sort_key_prefix = f"rsspost/{parent_document_id}/"
```
Comment (massi-ang) [inline lib/shared/layers/python-sdk/python/genai_core/documents.py:65]: Can we make this more generic?
```
if parent_document_id != None:
  sort_key_prefix = sort_key_prefix+f"{parent_document_id}/"
```

File: lib/shared/layers/python-sdk/python/genai_core/documents.py
Comment (massi-ang) [inline lib/shared/layers/python-sdk/python/genai_core/documents.py:1]: run `black` formatter on the file

File: lib/shared/layers/python-sdk/python/genai_core/documents.py [hunk 7]
```diff
@@ -190,9 +208,30 @@ def set_status(workspace_id: str, document_id: str, status: str):
             ":timestampValue": timestamp,
         },
     )
+    update_timestamp(workspace_id, document_id)
```
Comment (massi-ang) [inline lib/shared/layers/python-sdk/python/genai_core/documents.py:211]: the content of this function seems to be redundant with the above code. Was the idea to move the whole logic into the `update_timestamp` function?
the return value is not used, is this correct?

File: lib/rag-engines/rag-dynamodb-tables/index.ts
Comment (massi-ang) [inline lib/rag-engines/rag-dynamodb-tables/index.ts:1]: run `prettify` on this file

File: lib/shared/layers/python-sdk/python/genai_core/opensearch/delete.py
Comment (massi-ang) [inline lib/shared/layers/python-sdk/python/genai_core/opensearch/delete.py:1]: revert this change

File: lib/shared/layers/python-sdk/python/genai_core/documents.py [hunk 8]
```diff
@@ -190,9 +208,30 @@ def set_status(workspace_id: str, document_id: str, status: str):
...
+    return response
 
+def update_timestamp(workspace_id: str, document_id: str):
+    timestamp = datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%S.%fZ")
```
Comment (massi-ang) [inline lib/shared/layers/python-sdk/python/genai_core/documents.py:215]: consider creating a function to return the DDB formatted timestamp since this is used in several places

File: lib/shared/layers/python-sdk/python/genai_core/documents.py [hunk 9]
```diff
@@ -500,3 +567,158 @@ def _upload_document_content(
...
+                    "compound_sort_key": "rsspost/" + document_id + "/" + post_id,
+                    "status": "pending",
+                    "title": feed_entry['title'],
+                    "path": feed_entry['link'],
+                    "size_in_bytes": 0,
+                    "vectors": 0,
+                    "errors": [],
+                    "created_at": timestamp,
+                    "updated_at": timestamp,
+                    "crawler_properties": rss_document['crawler_properties'] if 'crawler_properties' in rss_document else None,
```
Comment (massi-ang) [inline lib/shared/layers/python-sdk/python/genai_core/documents.py:671]: replace with `dict.get("key", default)` expression

File: lib/shared/layers/python-sdk/python/genai_core/documents.py [hunk 10]
```diff
@@ -500,3 +567,158 @@ def _upload_document_content(
...
+                post_id = str(_get_hash_id_from_path(feed_entry['link']))
+                document = {
+                    "format_version": 1,
+                    "workspace_id": workspace_id,
+                    "document_id": post_id,
+                    "rss_feed_id": document_id,
+                    "document_type": "rsspost",
+                    "document_sub_type": None,
+                    "sub_documents": None,
+                    "compound_sort_key": "rsspost/" + document_id + "/" + post_id,
```
Comment (massi-ang) [inline lib/shared/layers/python-sdk/python/genai_core/documents.py:662]: maybe replace with `f"rsspost/{document_id}/{post_id}`

File: lib/user-interface/react-app/src/common/api-client/api-client.ts
Comment (massi-ang) [inline lib/user-interface/react-app/src/common/api-client/api-client.ts:1]: revert these changes

File: lib/user-interface/react-app/src/common/api-client/documents-client.ts
Comment (massi-ang) [inline lib/user-interface/react-app/src/common/api-client/documents-client.ts:1]: run `prettify` on this file

File: lib/user-interface/react-app/src/common/types.ts [hunk 11]
```diff
@@ -137,8 +138,26 @@ export interface DocumentItem {
   subDocuments: number;
   createdAt: string;
   updatedAt: string;
+  rssFeedId?: string;
+  rssLastCheckedAt?: string;
+  crawlerProperties?: {
+    followLinks: boolean;
+    limit: number;
+  }
 }
```
Comment (massi-ang) [inline lib/user-interface/react-app/src/common/types.ts:147]: can you explain why this information, which is related to the source of the document, must be included in the `DocumentItem` type?

File: lib/user-interface/react-app/src/pages/chatbot/playground/multi-chat-playground.tsx
Comment (massi-ang) [inline lib/user-interface/react-app/src/pages/chatbot/playground/multi-chat-playground.tsx:1]: revert this change or run `prettify` on this file

File: lib/user-interface/react-app/src/pages/rag/workspace/documents-tab.tsx
Comment (massi-ang) [inline lib/user-interface/react-app/src/pages/rag/workspace/documents-tab.tsx:1]: run `prettify` on this file

File: lib/user-interface/react-app/src/pages/rag/workspace/documents-tab.tsx [hunk 12]
```diff
@@ -101,12 +104,12 @@ export default function DocumentsTab(props: DocumentsTabProps) {
   const typeStr = ragDocumentTypeToString(props.documentType);
   const typeAddStr = ragDocumentTypeToAddString(props.documentType);
   const typeTitleStr = ragDocumentTypeToTitleString(props.documentType);
-
+  const columnDefinitions = getColumnDefinition(props.documentType);
   return (
```
Comment (massi-ang) [inline lib/user-interface/react-app/src/pages/rag/workspace/documents-tab.tsx:108]: add new line in between

File: lib/user-interface/react-app/src/pages/rag/workspace/rss-feed.tsx
Comment (massi-ang) [inline lib/user-interface/react-app/src/pages/rag/workspace/rss-feed.tsx:1]: run `prettify` on this file

File: package.json (hunk 1 above)
Comment (massi-ang) [reply package.json]: If you need that specific version because the feature you are using is only available there, ensure you bump the versions of all the CDK packages.

**Review by massi-ang — CHANGES_REQUESTED** [review]

File: lib/rag-engines/data-import/website-crawling-workflow.ts
Comment (flamingquaks) [reply lib/rag-engines/data-import/website-crawling-workflow.ts:1]: @massi-ang - This was something I wasn't 100% sure about. 
It's an extension of website crawling, but I can shift into an isolated construct if it would be better

File: lib/chatbot-api/functions/api-handler/routes/documents.py [hunk 13]
```diff
@@ -156,20 +229,73 @@ def add_document(workspace_id: str, document_type: str):
...
+            )
+            return {
+                "ok": True,
+                "data": "done"
+            }
+
 
 
 def _convert_document(document: dict):
+    if "crawler_properties" in document:
```
[the remark this answers is not in the document]
Comment (flamingquaks) [reply lib/chatbot-api/functions/api-handler/routes/documents.py:279]: @massi-ang I refactored this with the latest push to be a bit simpler. Basically the `crawler_properties` dict needs to format the objects within to have proper keys

File: lib/shared/layers/python-sdk/python/genai_core/documents.py (hunk 7 above)
Comment (flamingquaks) [reply lib/shared/layers/python-sdk/python/genai_core/documents.py]: You are correct. I removed `update_timestamp` function. The `update_subscription_timestamp` still exists and is leveraged throughout the code.

File: lib/user-interface/react-app/src/common/types.ts (hunk 11 above)
Comment (flamingquaks) [reply lib/user-interface/react-app/src/common/types.ts]: On the `rss-feed.tsx` page, both the RSS Feed subscription and the RSS Feed posts display this information.
For the Feed, it's showing the currently configured crawler properties that will be leveraged by website crawling requests triggered by detecting new feed posts. 
On the individual feed records, it shows what was configured when the feed post was ingested and sent for crawling. 
sample

**Comment by flamingquaks** [comment]
@massi-ang, I moved RSS Subscription to its own construct.

**Review by massi-ang — APPROVED** [review]

File: lib/shared/layers/python-sdk/python/genai_core/opensearch/delete.py
[the remark this answers is not in the document]
Comment (flamingquaks) [reply lib/shared/layers/python-sdk/python/genai_core/opensearch/delete.py:1]: @massi-ang - This appears in the PR as there was originally functionality in the workspace deletion, but has been removed. I don't know how to remove this blank change.

**Review by bigadsoleiman — APPROVED** [review]
